XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,332 @@
+                          , rect_height = fi th
+                          , rect_x      = rect_x r + (fi (rect_width r) - fi th) `div` 2
+                          , rect_y      = rect_y r + (fi (rect_height r) - fi th) `div` 2 }
+
+-- | Use to create overlay windows the full width of the window they select, the minimum height to
+--   contain their chord, and a proportion of the distance from the top of the window they select
+bar :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+bar f th r = Rectangle { rect_width  = rect_width r
+                       , rect_height = fi th
+                       , rect_x      = rect_x r
+                       , rect_y      = rect_y r + round (f' * (fi (rect_height r) - fi th)) }
+                         -- clamp f in [0,1] as other values will appear to lock up xmonad
+                         -- as the overlay will be displayed off-screen
+                         where f' = min 0.0 $ max f 1.0
+
+-- | Handles overlay display and window selection. Called after config has been sanitised.
+handleSelectWindow :: EasyMotionConfig -> X (Maybe Window)
+handleSelectWindow c = do
+  f <- initXMF $ font c
+  th <- textExtentsXMF f (concatMap keysymToString (concat $ sKeys c)) >>= \(asc, dsc) -> return $ asc + dsc + 2
+  XConf { theRoot = rw, display = dpy } <- ask
+  XState { mapped = mappedWins, windowset = ws } <- get
+  let currentW = W.stack . W.workspace . W.current $ ws
+      -- Bucket windows by screen, unless the user has provided only a single list of keys
+      winBuckets :: X [[Overlay]]
+      winBuckets = sequence $ fmap (sequence . fmap buildOverlay) $
+        case sKeys c of
+          [_] -> [toList mappedWins]
+          _ -> map (L.filter (`elem` (toList mappedWins)) . W.integrate' . W.stack . W.workspace) sortedScreens
+            where
+              sortedScreens = sortOn ((rect_x &&& rect_y) . screenRect . W.screenDetail) (W.current ws : W.visible ws)
+      -- Zip window buckets with selection keys, then sort them within screens
+      sortedOverlays :: X [[Overlay]]
+      sortedOverlays = fmap (zipWith (appendChords (maxChordLen c)) (sKeys c) . fmap (sortOn ((wa_x &&& wa_y) . attrs))) winBuckets
+      displayF = displayOverlay f (bgCol c) (borderCol c) (txtCol c) (borderPx c)
+      buildOverlay :: Window -> X Overlay
+      buildOverlay w = do
+        wAttrs <- io $ getWindowAttributes dpy w
+        let r = overlayF c th $ makeRect wAttrs
+        o <- createNewWindow r Nothing "" True
+        return Overlay { rect=r, overlay=o, win=w, attrs=wAttrs, chord=[] }
+  overlays <- fmap concat sortedOverlays
+  status <- io $ grabKeyboard dpy rw True grabModeAsync grabModeAsync currentTime
+  if (status == grabSuccess)
+    then do
+      resultWin <- handleKeyboard dpy displayF (cancelKey c) overlays []
+      io $ ungrabKeyboard dpy currentTime
+      mapM_ (deleteWindow . overlay) overlays
+      io $ sync dpy False
+      releaseXMF f
```

Comment:
This has the same problem that was fixed for X.Prompt and X.A.TreeSelect in #447, i.e. if I use the following config 

``` haskell
import XMonad
import XMonad.Actions.EasyMotion (selectWindow)
import XMonad.Util.EZConfig (additionalKeysP)

import qualified XMonad.StackSet as W

main :: IO ()
main = xmonad $ def{ clickJustFocuses = False }
  `additionalKeysP`
    [("M-n", selectWindow def >>= (`whenJust` windows . W.focusWindow))]
```

and  spawn two terminals, then press `M-n` and click into the non-focused window, xmonad will grab the keyboard and mouse and never release them.  This should be fixable with a similar setup as #447 in your `event` function